src/algebraic_geometry/morphisms/basic.lean

```diff
@@ -0,0 +1,152 @@
+/-
+Copyright (c) 2022 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+import algebraic_geometry.AffineScheme
+import algebraic_geometry.pullbacks
+
+/-!
+# Properties of morphisms
+
+We provide the basic framework for talking about properties of morphisms.
+The following meta-properties are defined
+
+* `respects_iso`: `P` respects isomorphisms if `P f → P (e ≫ f)` and `P f → P (f ≫ e)`, where
+  `e` is an isomorphism.
+* `stable_under_composition`: `P` is stable under composition if `P f → P g → P (f ≫ g)`.
+* `stable_under_base_change`: `P` is stable under base change if `P (Y ⟶ S) → P (X ×[S] Y ⟶ X)`.
+
+-/
+
+universe u
+
+open topological_space category_theory category_theory.limits opposite
+
+noncomputable theory
+
+namespace algebraic_geometry
+
+/-- A `morphism_property` is a class of morphisms between schemes. -/
+@[derive complete_lattice]
+def morphism_property := ∀ ⦃X Y : Scheme⦄ (f : X ⟶ Y), Prop
+
+/-- An `affine_target_morphism_property` is a class of morphisms from an arbitrary scheme into an
+affine scheme. -/
+def affine_target_morphism_property := ∀ ⦃X Y : Scheme⦄ (f : X ⟶ Y) [is_affine Y], Prop
+
+/-- `is_iso` as a `morphism_property`. -/
+protected def Scheme.is_iso : morphism_property := @is_iso Scheme _
+
+/-- `is_iso` as an `affine_morphism_property`. -/
+protected def Scheme.affine_target_is_iso : affine_target_morphism_property :=
+λ X Y f H, is_iso f
+
+instance : inhabited morphism_property := ⟨Scheme.is_iso⟩
+instance : inhabited affine_target_morphism_property := ⟨Scheme.affine_target_is_iso⟩
+
+/-- A `affine_target_morphism_property` can be extended to a `morphism_property` such that it
+*never* holds when the target is not affine -/
+def affine_target_morphism_property.to_property (P : affine_target_morphism_property) :
+  morphism_property :=
+λ X Y f, ∃ h, @@P f h
+
+lemma affine_target_morphism_property.to_property_apply (P : affine_target_morphism_property)
+  {X Y : Scheme} (f : X ⟶ Y) [is_affine Y] :
+  P.to_property f ↔ P f := by { delta affine_target_morphism_property.to_property, simp [*] }
+
+/-- A morphism property `respects_iso` if it still holds when composed with an isomorphism -/
+def respects_iso (P : morphism_property) :=
```

Comment:
Since you are sorta using `Prop` on multiple levels, I thought it might be useful to include the type here, for readers just glancing over this file. Also elsewhere.
```suggestion
def respects_iso (P : morphism_property) : Prop :=
```